File: app/src/main/java/com/ipd/taxiu/ui/fragment/collect/CollectStoreFragment.java
package com.ipd.taxiu.ui.fragment.collect;

import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;

import com.ipd.taxiu.R;
import com.ipd.taxiu.adapter.CollectStoreAdapter;
import com.ipd.taxiu.bean.CollectStoreBean;
import com.ipd.taxiu.bean.CollectStoreListBean;
import com.ipd.taxiu.ui.ListFragment;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;

import rx.Observable;
import rx.Subscriber;

/**
 * Created by Miss on 2018/7/19
 */
public class CollectStoreFragment extends ListFragment<CollectStoreListBean, CollectStoreBean> {
    private CollectStoreAdapter mAdapter = null;

    public static CollectStoreFragment newInstance() {
        CollectStoreFragment fragment = new CollectStoreFragment();
        return fragment;
    }

    @Override
    protected void initView(@Nullable Bundle bundle) {
        super.initView(bundle);
        progress_layout.setEmptyViewRes(R.layout.layout_empty_collect);
    }

    @NotNull
    @Override
    public Observable<CollectStoreListBean> loadListData() {
        return Observable.create(new Observable.OnSubscribe<CollectStoreListBean>() {
            @Override
            public void call(Subscriber<? super CollectStoreListBean> subscriber) {
                CollectStoreListBean bean = new CollectStoreListBean();
                bean.list = new ArrayList<>();
                for (int i = 0; i < 2; i++) {
                    bean.list.add(new CollectStoreBean());
                }
                subscriber.onNext(bean);
                subscriber.onCompleted();
            }
        });
    }

    @Override
    public int isNoMoreData(CollectStoreListBean result) {
        if (result.list == null || result.list.isEmpty()) {
            return getEMPTY_DATA();
        } else {
            return getNORMAL();
        }
    }

    @Override
    public void setOrNotifyAdapter() {
        if (mAdapter == null) {
            mAdapter = new CollectStoreAdapter(getContext(), getData());
            recycler_view.setLayoutManager(new GridLayoutManager(getContext(), 2));
            recycler_view.setAdapter(mAdapter);
        } else {
            mAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void addData(boolean isRefresh, CollectStoreListBean result) {
        getData().addAll(result.list);
    }
}

File: app/src/main/java/com/ipd/taxiu/ui/fragment/collect/CollectStoreFragment.kt
package com.ipd.taxiu.ui.fragment.classroom

import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import com.ipd.taxiu.R
import com.ipd.taxiu.adapter.ProductAdapter
import com.ipd.taxiu.bean.BaseResult
import com.ipd.taxiu.bean.ProductBean
import com.ipd.taxiu.event.UpdateCollectProductEvent
import com.ipd.taxiu.platform.global.Constant
import com.ipd.taxiu.platform.global.GlobalParam
import com.ipd.taxiu.platform.http.ApiManager
import com.ipd.taxiu.ui.ListFragment
import com.ipd.taxiu.ui.activity.store.ProductDetailActivity
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import rx.Observable

class CollectStoreFragment : ListFragment<BaseResult<List<ProductBean>>, ProductBean>() {
    companion object {
        fun newInstance(): CollectStoreFragment {
            return CollectStoreFragment()
        }
    }

    override fun onViewAttach() {
        super.onViewAttach()
        EventBus.getDefault().register(this)
    }

    override fun onViewDetach() {
        super.onViewDetach()
        EventBus.getDefault().unregister(this)
    }

    override fun needLazyLoad() = true


    override fun initView(bundle: Bundle?) {
        super.initView(bundle)
        progress_layout.setEmptyViewRes(R.layout.empty_base)
    }

    override fun loadListData(): Observable<BaseResult<List<ProductBean>>> {
        return ApiManager.getService().productCollect(GlobalParam.getUserIdOrJump(), Constant.PAGE_SIZE, page)
    }

    override fun isNoMoreData(result: BaseResult<List<ProductBean>>): Int {
        if (page == INIT_PAGE && (result.data == null || result.data.isEmpty())) {
            return EMPTY_DATA
        } else if (result.data == null || result.data.isEmpty()) {
            return NO_MORE_DATA
        }
        return NORMAL
    }

    private var mAdapter: ProductAdapter? = null
    override fun setOrNotifyAdapter() {
        if (mAdapter == null) {
            mAdapter = ProductAdapter(mActivity, data, {
                //itemClick
                ProductDetailActivity.launch(mActivity, it.PRODUCT_ID, it.FORM_ID)
            })
            mAdapter?.switchShowType()
            recycler_view.layoutManager = GridLayoutManager(mActivity, 2)
            recycler_view.adapter = mAdapter
        } else {
            mAdapter?.notifyDataSetChanged()
        }
    }

    override fun addData(isRefresh: Boolean, result: BaseResult<List<ProductBean>>) {
        data?.addAll(result?.data ?: arrayListOf())
    }

    @Subscribe
    fun onMainEvent(event: UpdateCollectProductEvent) {
        if (isFirstLoad()) return
        onRefresh(true)
    }

}